policies/coding-style.md

```diff
@@ -586,11 +586,64 @@ it portable enough to meet our multi-platform support goals.
 
 ## Chapter 14:  Portability
 
+### Language
+
 To maximise portability the version of C defined in ISO/IEC 9899:1990
 should be used. This is more commonly referred to as C90. ISO/IEC 9899:1999
 (also known as C99) is not supported on some platforms that OpenSSL is
 used on and therefore should be avoided.
 
+### Integer types
+
+The types `int8_t`, `uint8_t`, `int16_t`, `uint16_t`, `int32_t`, `uint32_t`,
+`int64_t` and `uint64_t` are recommended for use when the exact width is
+important. These were only standardised in C99 but we define them ourselves if
+they are not available on a target platform.
+
+`size_t` is available. `ossl_ssize_t` should be used instead of `ssize_t` for
+portability.
+
+Most of the built-in integer types in the C language should be avoided
+or have important caveats:
+
+- Use of `long` and `unsigned long` is deprecated, as these types cannot be
```

Comment:
The _exact_ same arguments for not using `long` also apply to `long long`.  It's is only guaranteed to be as wide as an `int`.

Reply:
This is debatable. IMO we depend on long long being 64 bit and that is held on all platforms we currently support. Of course we could replace long long with int/uint64_t except with BIO_snprintf it is a little bit awkward to use `%lld` with cast to `(int64_t)`. So maybe we should deprecate long long and unsigned long long except for use with printf.

Reply:
Existing uses have to remain of course, but I stand by my statement that the **exact** same arguments against `long` also apply to `long long`.  If we're deprecating the former, we should also deprecate the latter.  Regardless of the problems it might cause.

Yes, `BIO_printf` et al are problematic, yet unmentioned here.  They have zero scope for supporting the fixed sized integers.  Their supporting `size_t` was (mildly) annoying when the `z` modifier was added.

Reply:
We should probably use the PRI macros a bit more consistently

https://en.cppreference.com/w/c/types/integer

Reply:
This would remove my objection to deprecating `short int`.

They aren't in C89 and we'd need to implement wrappers if we don't accept C99.

Reply:
I actually think we should start small and just deprecate long and unsigned long, for others we just state the current limits we depend on in the code base.

Reply:
Nah, I'd rather not do this policy change incrementally.  I think it's safe enough if we have the (stated?) expectation that this applies primarily to new code, leaving the option of modifying old code exactly that, an option.

Reply:
I just checked and long long has to be at least 64 bits in both C99 and C11. But there's no reason to use it directly, so let's deprecate it.

The correct way to use fixed width integer types with printf is with the macros defined in inttypes.h:

```c
printf("Value: " PRIu32 "\n", (uint32_t)42);
```

This is admittedly painful. Though I think we may have to go here anyway as I've encounted a great volume of warnings when targeting mingw about not supporting `%zu`, which we use extensively.

Reply:
> This is admittedly painful. Though I think we may have to go here anyway as I've encounted a great volume of warnings when targeting mingw about not supporting `%zu`, which we use extensively.

It's only very recently that MSVC *stopped* giving those same warnings...

Reply:
We're using C89 currently....

Reply:
> We're using C89 currently....

... which is clear with our already liberal use of [u]int*nn*_t :wink:

Reply:
long long doesn't exist in C89.